sdk/eventhub/azure-eventhub/azure/eventhub/_utils.py

```diff
@@ -245,12 +258,84 @@ def get_last_enqueued_event_properties(event_data):
     return (sas, expiry)
+
+
+def transform_single_message_if_needed(message, message_type):
+    # type: (Union[AmqpAnnotatedMessage, EventData], Type[EventData]) -> EventData
+    """
+    This method serves multiple goals:
+    1. update the internal message to reflect any updates to settable properties on EventData
+    2. transform the AmqpAnnotatedMessage to be EventData
+    :param message: A single instance of message of type EventData
+        or AmqpAnnotatedMessage.
+    :type message: ~azure.eventhub.common.EventData, ~azure.eventhub.amqp.AmqpAnnotatedMessage
+    :param Type[EventData] message_type: The class type to return the messages as.
+    :rtype: EventData
+    """
+    try:
+        # EventData
+        # pylint: disable=protected-access
+        return message._to_outgoing_message()  # type: ignore
+    except AttributeError:
+        # AmqpAnnotatedMessage
+        # pylint: disable=protected-access
+        return message_type._from_message(
+            message=message._to_outgoing_amqp_message(), raw_amqp_message=message  # type: ignore
+        )
+
+
+def transform_messages_if_needed(messages, message_type):
+    # type: (MessagesType, Type[EventData]) -> Union[EventData, List[EventData]]
+    """
+    This method serves multiple goals:
+    1. update the internal messages to reflect any updates to settable properties on EventData
+    2. transform the AmqpAnnotatedMessage to be EventData
+    :param MessagesType messages: A list or single instance of messages of type EventData 
+        or AmqpAnnotatedMessage.
+    :param Type[EventData] message_type: The class type to return the messages as.
+    :rtype: Union[EventData, List[EventData]]
+    """
+    if isinstance(messages, Iterable):
+        return [transform_single_message_if_needed(m, message_type) for m in messages]
+    return transform_single_message_if_needed(messages, message_type)
+
+
+def decode_with_recurse(data, encoding="UTF-8"):
+    # type: (Any, str) -> Any
+    """
+    If data is of a compatible type, iterates through nested structure and decodes all binary
+        strings with provided encoding.
+    :param Any data: The data object which, if compatible, will be iterated through to decode binary string.
+    :param encoding: The encoding to use for decoding data.
+        Default is 'UTF-8'
+    :rtype: Any
+    """
+
+    if isinstance(data, str):
+        return data
+    if isinstance(data, six.binary_type):
+        return data.decode(encoding)
+    if isinstance(data, Mapping):
+        decoded_mapping = {}
+        for k,v in data.items():
+            decoded_key = decode_with_recurse(k, encoding)
+            decoded_val = decode_with_recurse(v, encoding)
+            decoded_mapping[decoded_key] = decoded_val
+        return decoded_mapping
+    if isinstance(data, Iterable):
+        decoded_list = []
+        for d in data:
+            decoded_list.append(decode_with_recurse(d, encoding))
+        return decoded_list
```

Comment:
I'm thinking of cases when data of type integer, float, boolean, etc. is passed but it looks they are not handled.
probably we just need to handle it like this?
```python
if isinstance(data, Mapping):
    #recursive
elif isinstance(data, Iterable):
    #recursive
elif if isinstance(data, str):

else:
    return data
```

Reply:
so for integer, float, boolean, we should just be returning them as is. the thing that you did above will result in recursion error b/c string is actually `Iterable`. If instead of what you said above, we did:
```
if isinstance(data, str):
    return data
elif isinstance(data, Mapping):
    #recursive
elif isinstance(data, Iterable):
    #recursive
elif if isinstance(data, six.binary_type):
    return data.decode(encoding)
return str(data)
```
This would result in the types that you mentioned being turned to string before returning. So passing in `"[1, True, {b'key': ['hi', 4.5]}]"` would result in `['1', 'True', {'key': ['hi', '4.5']}]`. This probably isn't what the user would be expecting. By returning `data` if it doesn't fall under any of the conditions, and calling `str(data)` on the returned object in `_decode_non_body_data()`, this would result in `"[1, True, {'key': ['hi', 4.5]}]"`.